scalyr_agent/connection.py

```diff
@@ -445,6 +445,63 @@ def connect(self):
         if hasattr(self, "_tunnel_host") and self._tunnel_host:
             self._tunnel()
 
+        def log_server_ssl_certificate(py_post_equal_279):
+            # type: (bool) -> None
+            """
+            Utility method which establishes new connection to the server using the same parameters
+            as the original parameter and logs the server certificate in PEM format.
+
+            NOTE: This method intentionally doesn't perform certificate validation - we just want to
+            retrieve raw certificate and log it to make it easier to troubleshoot certificate
+            validation failure related issues.
+            """
+            # NOTE: We can't use ssl.get_server_certificate() since this will establish a new
+            # connection and won't use the same parameters as the original connection (it also won't
+            # send SNI so a different cert may be returned).
+            # We also can't use sock.getpeercert() directly since for the original non-logging
+            # socket, ca validation is turned on (as it should always be!) and it will throw in
+            # case of invalid cert - this method is here purely for logging purposes so we can log
+            # problematic cert on failed cert validation so for logging purposes we don't want to
+            # perform any cert validation.
+            # It's worth noting that here we can't re-use original sock either and need to create a
+            # new connection - not ideal since a different cert may be returned for a new
+            # connection, but at least SNI is used and we use the same parameters as we do for
+            # normal non-logging connection.
+            sock = None
+
+            try:
+                sock = socket.create_connection((self.host, self.port), self.__timeout)
+
+                if py_post_equal_279:
+                    ssl_context = ssl.SSLContext(ssl.PROTOCOL_TLSv1_2)
+                    ssl_context.options |= ssl.OP_NO_SSLv2
+                    ssl_context.options |= ssl.OP_NO_SSLv3
+                    ssl_context.options |= ssl.OP_NO_TLSv1
+                    ssl_context.options |= ssl.OP_NO_TLSv1_1
+                    ssl_context.verify_mode = ssl.CERT_NONE
+                    ssl_context.check_hostname = False
+
+                    sock = ssl_context.wrap_socket(
+                        sock, do_handshake_on_connect=True, server_hostname=self.host
+                    )
+
+                else:
+                    sock = ssl.wrap_socket(sock, cert_reqs=ssl.CERT_NONE)
+
+                # NOTE: In case binary_form is False and validation wasn't perform an empty dict
+                # will be returned so we need to use binary_form to still get cert back for logging
+                # purposes - https://docs.python.org/3/library/ssl.html#ssl.SSLSocket.getpeercert
+                der_cert = sock.getpeercert(binary_form=True)
```

Comment:
Shouldn't we guard it from possible ValueError exception, which is mentioned in docs?

Reply:
oops, sorry, yeah, good catch - i thought I also added ``except`` clause, but I only added ``finally``. Will fix it and also add ``except Exception``.